prometheus/src/main/java/com/uber/m3/tally/prometheus/PrometheusReporter.java

```diff
@@ -0,0 +1,442 @@
+                .name(name)
+                .help(String.format("%s counter", name))
+                .labelNames(collectionToStringArray(ttags.keySet()))
+                .register(registry));
+        registeredCounters.get(collectorName).labels(collectionToStringArray(ttags.values())).inc(value);
+    }
+
+    @Override
+    public void reportGauge(String name, Map<String, String> tags, double value) {
+        final Map<String, String> ttags = (tags == null) ? Collections.emptyMap() : tags;
+        String collectorName = canonicalMetricId(name, ttags.keySet());
+        registeredGauges.computeIfAbsent(collectorName, key -> Gauge.build()
+                .name(name)
+                .help(String.format("%s gauge", name))
+                .labelNames(collectionToStringArray(ttags.keySet()))
+                .register(registry));
+        registeredGauges.get(collectorName).labels(collectionToStringArray(ttags.values())).set(value);
+    }
+
+    @Override
+    public void reportTimer(String name, Map<String, String> tags, Duration interval) {
+        switch (timerType) {
+            case HISTOGRAM:
+                reportTimerHistogram(name, tags, interval);
+                break;
+            case SUMMARY:
+            default:
+                reportTimerSummary(name, tags, interval);
+        }
+    }
+
+    @Override
+    public void reportHistogramValueSamples(
+            String name,
+            Map<String, String> tags,
+            Buckets buckets,
+            double bucketLowerBound,
+            double bucketUpperBound,
+            long samples
+    ) {
+        final Map<String, String> ttags = (tags == null) ? Collections.emptyMap() : tags;
+        String collectorName = canonicalMetricId(name, ttags.keySet());
+        registeredHistograms.computeIfAbsent(collectorName, key -> {
+            double[] b = buckets.getValueUpperBounds().stream().mapToDouble(a -> a).toArray();
+            return Histogram.build()
+                    .name(name)
+                    .help(String.format("%s histogram", name))
+                    .buckets(b)
+                    .labelNames(collectionToStringArray(ttags.keySet()))
+                    .register(registry);
+        });
+        Histogram.Child histogram = registeredHistograms.get(collectorName)
+                .labels(collectionToStringArray(ttags.values()));
+        for (int i = 0; i < samples; i++) {
+            histogram.observe(bucketUpperBound);
+        }
+    }
+
+    @Override
+    public void reportHistogramDurationSamples(
+            String name,
+            Map<String, String> tags,
+            Buckets buckets,
+            Duration bucketLowerBound,
+            Duration bucketUpperBound,
+            long samples
+    ) {
+        final Map<String, String> ttags = (tags == null) ? Collections.emptyMap() : tags;
+        String collectorName = canonicalMetricId(name, ttags.keySet());
+        registeredHistograms.computeIfAbsent(collectorName, key -> {
```

Comment:
Hmm so one problem I didn't anticipate with `computeIfAbsent`: it looks like the implementation can call the lambda multiple times on conflicts:

```
The default implementation may retry these steps when multiple threads attempt updates including potentially calling the mapping function multiple times.
```

which can cause problems with `register` (io.prometheus.client.CollectorRegistry#register)--it throws an `IllegalArgumentException` on conflicts:

```
public void register(Collector m) {
    List<String> names = collectorNames(m);
    synchronized (namesCollectorsLock) {
      for (String name : names) {
        if (namesToCollectors.containsKey(name)) {
          throw new IllegalArgumentException("Collector already registered that provides name: " + name);
        }
      }
      ...
  }
```

@prateek @SokolAndrey thoughts on the right approach here? We could take the simple approach, and just throw the get/compute flow behind a synchronized block ourselves instead, e.g.:

```
Histogram.Child getOrCreateHisto(String key) {
  // quick check first
  Histogram maybeHisto = registeredHistograms.get(key);
  if (maybeHisto == null) {
     return registeredHistograms.get(key);
  }

   // acquire the lock to make sure noone else is trying to create this key. Bit heavy; we're 
   // locking all histogram metric creation here. Prom also global locks--we could rely on that
   // instead?
   
     synchronized (registeredHistograms) {
        // create the histogram if needed
        Histogram maybeHisto = registeredHistograms.get(key);
        // maybe someone else beat us here
        if (maybeHisto != null) {
             return maybeHisto
        }
        histo = new Histogram()...register(registry);
        registeredHistograms.put(key, histo);
        // we're the only one's trying to create this 
     }
  }
}
```

Not sure whether the above is correct/the most efficient way of doing things; just an idea to start with. We could also do that same synchronization within the `computeIfAbsent` block; dunno if that's more or less clear.